### PR TITLE
Numerics/integer

### PR DESCRIPTION
**Pull request by bennn**
Not ready to merge, but up for style review

### CODE REVIEW

File: main.rkt [hunk 1]
```diff
@@ -22,6 +25,9 @@
 
 ;; =============================================================================
 
+(define (square n)
+  (expt n 2))
+
```
Comment (bkc39) [inline main.rkt]: Amazing that this isn't here. For the sake of all the power series stuff, we should add

``` racket
(define (cube n)
  (expt n 3))
```

In my file I also defined some useful constants

``` racket
(define π pi)
(define π/2 (/ π 2))
(define π/3 (/ π 3))
(define π/4 (/ π 4))
(define π/6 (/ π 6))
(define π/12 (/ π 12))
(define 2/π (/ 2 π))
(define 3π/4 (* 3 π/4))
(define e (exp 0))
(define 1/e (/ 1 e))
```

These are probably best added here for now, then when we amass enough of them, we should export a `constants.rkt` and just import from that.

Comment (bkc39) [inline main.rkt]: Also hilariously absent is `fact`. The question is how do we want to implement it..

```
(define (fact n)
  (let loop ([a 0] [n n])
    (if (= n 0)
        1
        (loop (* n a) (sub1 n)))))

(require math/special-functions)
(define fact (compose truncate gamma add1))
```

Also for naming do we prefer `fact` or `factorial`? The first is probably more efficient, but the second one doesn't require integer arguments.. Maybe export both and have them have different contracts?

Comment (bennn) [inline main.rkt]: I think we should use Racket's `factorial`. (I'd be happy if some day, lots of these functions from `scmutils` get absorbed into the Racket math library.)
http://docs.racket-lang.org/math/number-theory.html?q=fact#%28def._%28%28lib._math%2Fnumber-theory..rkt%29._factorial%29%29

Comment (bkc39) [inline main.rkt]: Wow.. I searched and couldn't find it. Why the hell is it under "number theory"? But yea if its already there let's use that. I'd also like to export unicode synonyms for most of the constants and special functions like gamma, psi, von-mangoldt, moebius, etc

**Comment by bennn** [comment]
the SICM code uses the rational `(make-rational 1 0)`. I asked how to get a rational number in Racket such that `(numerator x)` is 1 and `(denominator x)` is 0, but it's not possible.

We'd have to make our own rational numbers and I am not interested in that.

**Comment by bkc39** [comment]
How do they use `(make-rational 1 0)`? Their implementation probably just has an extended for the extended rationals. If the expected behavior is infinity then we could just patch with

``` racket
(define (extended-rational numerator denominator)
  (if denominator
      (/ numerator denominator)
      (if (> numerator 0) +inf.0 -inf.0)))
```

and just sub that in when they do division that shouldn't work.

**Comment by bkc39** [comment]
I propose to also export common fractions. It is tedious to keep writing `(/ 1 3)`. I propose

``` racket
(define 1/2 (/ 1 2))
(define 1/3 (/ 1 3))
(define 2/3 (/ 2 3))
(define 1/4 (/ 1 4))
(define 3/4 (/ 3 4))
(define 1/5 (/ 1 5))
(define 1/6 (/ 1 6))
(define 1/8 (/ 1 8))
(define 1/10 (/ 1 10))
(define 1/12 (/ 1 12))
```

I'd be down to add more but these have all shown up repeatedly for me nat this point.

**Comment by bkc39** [comment]
Huh.. looks like you can't define those. Never mind.

**Comment by bennn** [comment]
Lol yeah those are all literal rational numbers. Just like saying `(define 3 4)`

**Comment by bkc39** [comment]
Wait so you can literally write `1/6` instead of `(/ 1 6)`? I'm not in front of an interpreter right now.

**Comment by bennn** [comment]
Yes. I think this is really funny.

```
Welcome to Racket v6.2.0.2.
-> (+ 1 1/6)
7/6
```

**Comment by bkc39** [comment]
That's dope

File: main.rkt [hunk 2]
```diff
@@ -8,6 +8,25 @@
...
+ π/3
+ π/4
+ π/6
+ π/12
+ 2/π
+ 3π/4
+ e
+ 1/e
+ ;; --- helpers
+ square
```
Comment (bkc39) [inline main.rkt]: Bruh. `square` is in the library as [`sqr`](http://docs.racket-lang.org/reference/generic-numbers.html#%28def._%28%28lib._racket%2Fmath..rkt%29._sqr%29%29). LOL. I suppose this doesn't matter as the built in implementation is the same as ours. But I think that this indicates that it is more difficult to find the functions you need in the stdlib than it should be.

File: private/numerics/integer/continued-fractions.rkt [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+  (check-equal? (take (let ((number pi))
+                        (stream-map
+                         (lambda (c)
+                           (let ((error (abs (- c number))))
+                             (- error (/ 1 (expt (denominator c) 2)))))
+                         (convergents
+                          (continued-fraction number)))) 20)
+                (list -.8584073464102069 -1.9143673997956443e-2 -5.780016472316501e-6 -7.804790414797462e-5 -3.347326313231945e-10 -5.747961301800781e-10 -1.0502260270843243e-10 -7.179925756609443e-11 -5.4838338802403175e-12 -5.89896667061906e-12 -1.3644086097276364e-13 -3.1384653391315194e-13 -1.0925052623624082e-15 -1.6328867414587005e-16 -1.4760984937032654e-17 -2.718809562442266e-20 -2.499649407750906e-20 -6.51444233883359e-21 -1.154885932677522e-22 -6.760478215475832e-24))
+)
```
Comment (bkc39) [inline private/numerics/integer/continued-fractions.rkt:175]: Ummmmmm. 80 char plz?

File: main.rkt (hunk 2 above)
Comment (bennn) [inline main.rkt]: Argh!

File: private/numerics/integer/continued-fractions.rkt (hunk 3 above)
Comment (bennn) [inline private/numerics/integer/continued-fractions.rkt:175]: but it's only test cases.... ok I'll change it

**Comment by bennn** [comment]
The code is done. Do you want scribblings, or can I merge?

**Comment by bkc39** [comment]
I'd prefer scribblings before merge. In general I think the prereq for merge should be
- [x] implementation
- [ ] tests
- [x] scribblings

Check out #7. If we merge that in then you can just rebase on top of that and add another section.

**Comment by bennn** [comment]
K we have docs
screen shot 2015-10-10 at 1 16 45 am

**Comment by bennn** [comment]
(The bessel functions are the failure)

**Comment by bkc39** [comment]
LGTM

**Pull request merged**